### PR TITLE
[Dynamo] only import einops if version is lower than 0.7.0

### PR DESCRIPTION
**Pull request by mlazos**
Fixes internal xref (https://fb.workplace.com/groups/257735836456307/posts/804793021750583/?comment_id=805229281706957&reply_comment_id=805232695039949)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/decorators.py [hunk 1]
```diff
@@ -608,27 +608,37 @@ def mark_static_address(t, guard=True):
 # Note: this carefully avoids eagerly import einops.
 # TODO: we should delete this whole _allow_in_graph_einops logic by approximately 2024 Q2
 def _allow_in_graph_einops():
-    import einops
+    import sys
```
Comment (zou3519) [inline torch/_dynamo/decorators.py:611]: let's put this at the top lol

File: torch/_dynamo/decorators.py [hunk 2]
```diff
@@ -608,27 +608,37 @@ def mark_static_address(t, guard=True):
...
-    try:
-        # requires einops > 0.6.1, torch >= 2.0
-        from einops._torch_specific import (  # type: ignore[attr-defined]  # noqa: F401
-            _ops_were_registered_in_torchdynamo,
-        )
+    mod = sys.modules.get("einops")
+    if mod is None:
+        return
+    else:
+        from packaging import version
```
Comment (zou3519) [inline torch/_dynamo/decorators.py:617]: ditto here, lazy imports should only be used to resolve import cycles

**Review by zou3519 — APPROVED** [review]
sounds good

**Comment by atalman** [comment]
Seeing an error here:
https://github.com/pytorch/pytorch/actions/runs/12301892273/job/34335593226?pr=143125#step:15:1000
```
+ python -c 'from smoke_test import test_linalg; test_linalg()'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/pytorch/.ci/pytorch/smoke_test/smoke_test.py", line 11, in <module>
    import torch._dynamo
  File "/opt/python/cp311-cp311/lib/python3.11/site-packages/torch/_dynamo/__init__.py", line 8, in <module>
    from .decorators import (
  File "/opt/python/cp311-cp311/lib/python3.11/site-packages/torch/_dynamo/decorators.py", line 9, in <module>
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
```

**Comment by atalman** [comment]
Please reland this with ciflow/binaries label attached

File: torch/_dynamo/decorators.py [hunk 3]
```diff
@@ -2,9 +2,12 @@
 # ruff: noqa: TCH004
 import functools
 import inspect
+import sys
 from dataclasses import dataclass
 from typing import Any, Callable, Dict, Type, TYPE_CHECKING, TypeVar
 
+from packaging import version
```
Comment (malfet) [inline torch/_dynamo/decorators.py:9]: We should never used `packaging`, but instead use `torch_version` which tries to use packaging, and if not available has some sort of home-grown implementation

Comment (mlazos) [reply torch/_dynamo/decorators.py:9]: Sounds good, thanks!!

**Pull request closed without merging**